pkg/protocol/inactivity/states.go

```diff
@@ -0,0 +1,209 @@
+// claimSigningState is the state during which group members sign their
+// preferred inactivity claim (by hashing their inactivity, and then signing the
+// result), and share this over the broadcast channel.
+type claimSigningState struct {
+	*state.BaseAsyncState
+
+	channel        net.BroadcastChannel
+	claimSigner    ClaimSigner
+	claimSubmitter ClaimSubmitter
+
+	member *signingMember
+
+	claim *Claim
+}
+
+func (css *claimSigningState) Initiate(ctx context.Context) error {
+	message, err := css.member.signClaim(css.claim, css.claimSigner)
+	if err != nil {
+		return err
+	}
+
+	if err := css.channel.Send(
+		ctx,
+		message,
+		net.BackoffRetransmissionStrategy,
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (css *claimSigningState) Receive(netMessage net.Message) error {
+	// The network layer determines the message sender's public key based on
+	// the network client's pinned identity. The sender can not use any other
+	// public key than the one it is identified with in the network.
+	// Furthermore, the sender must possess the associated private key - each
+	// network message is signed with it.
+	//
+	// The network layer rejects any message with an incorrect signature or
+	// altered public key. By this point, we've conducted enough checks to
+	// be very certain that the sender' public key presented in the network
+	// net.Message is the correct one.
+	//
+	// In this final step, we compare the pinned network key with one used to
+	// produce a signature over the inactivity claim hash. If the keys don't
+	// match, it means that an incorrect key was used to sign inactivity claim
+	// hash and the message should be rejected.
+	isValidKeyUsed := func(signatureMessage *claimSignatureMessage) bool {
+		return bytes.Equal(signatureMessage.publicKey, netMessage.SenderPublicKey())
+	}
+
+	// As there is only one message type exchanged during result publication,
+	// we can simplify the code and cast directly to the concrete type
+	// `*resultSignatureMessage` instead of casting to the generic `message`.
+	if signatureMessage, ok := netMessage.Payload().(*claimSignatureMessage); ok {
+		if css.member.shouldAcceptMessage(
+			signatureMessage.SenderID(),
+			netMessage.SenderPublicKey(),
+		) && isValidKeyUsed(
+			signatureMessage,
+		) && css.member.sessionID == signatureMessage.sessionID {
+			css.ReceiveToHistory(netMessage)
+		}
+	}
+
+	return nil
+}
+
+func (css *claimSigningState) CanTransition() bool {
+	// Require the number of received signatures to be at least the honest
+	// threshold. Unlike in the case of DKG, we cannot expect all the members to
+	// participate in signing as we know we are dealing with some problem
+	// arising from operator inactivity.
+	// TODO: Consider passing the number of required signatures from the code
+	//       that launched the inactivity operator execution.
+	messagingDone := len(receivedMessages[*claimSignatureMessage](css.BaseAsyncState)) >=
+		css.member.group.HonestThreshold()
```

Comment:
Shouldn't we expect `>= css.member.group.HonestThreshold() - 1`?

Reply:
Yes, and it will always add their own signature on top. That means we need `threshold - 1` of 3rd party signatures to meet the required threshold.